### PR TITLE
KAFKA-16684: Remove cache in responseData

### PR DESCRIPTION
**Pull request by johnnychhsu**
## Context
The response data should change accordingly to the input, however with the current design, it will not change even if the input changes. We should remove this cache logic to avoid returning wrong data.
Jira: https://issues.apache.org/jira/browse/KAFKA-16684

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/common/requests/FetchResponse.java
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/requests/FetchResponse.java]: This is what I do concern! The `responseData` could be different if the input gets changed.

Comment (johnnychhsu) [reply clients/src/main/java/org/apache/kafka/common/requests/FetchResponse.java]: thanks for the review!
yes, the returned data should be calculated on the fly based on the input topic names

**Comment by chia7712** [comment]
@johnnychhsu This is a bug fix, so please add test for it

**Review by TaiJuWu — APPROVED** [review]
LGTM

**Review by apoorvmittal10 — APPROVED** [review]
LGTM! The only thing which I am not sure is that do we have any reason to cache the response for faster processing i.e. how frequently this method is called on same `topicNames`? If that call is too frequesnt with no change in `topicNames` then will it be worth to re-compute only if `topicNames` have changed?

The reason I say that is because the method call happens while fetching topicNames from session, which mostly will remain same.

**Comment by chia7712** [comment]
> The reason I say that is because the method call happens while fetching topicNames from session, which mostly will remain same.

that is a good point. Maybe the cache should be hosted by session instead of response data. the response data can be accessed by everyone, so it seems to me that is not a good choice to have cache.

**Comment by apoorvmittal10** [comment]
> > The reason I say that is because the method call happens while fetching topicNames from session, which mostly will remain same.
> 
> that is a good point. Maybe the cache should be hosted by session instead of response data. the response data can be accessed by everyone, so it seems to me that is not a good choice to have cache.

Sounds fair 👍

**Comment by chia7712** [comment]
> Sounds fair 

I file https://issues.apache.org/jira/browse/KAFKA-17065 to log it

**Comment by m1a2st** [comment]
LGTM, I wll add test after this PR merge into trunk.

**Review by m1a2st — APPROVED** [review]

**Comment by chia7712** [comment]
```
org.gradle.api.internal.tasks.testing.TestSuiteExecutionException: Could not complete execution for Gradle Test Executor 100.
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:64)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$3.run(TestWorker.java:193)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:119)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:66)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
Caused by: java.lang.OutOfMemoryError: GC overhead limit exceeded
	at net.bytebuddy.description.type.TypeDescription$ForLoadedType.of(TypeDescription.java:8619)
	at net.bytebuddy.description.method.MethodDescription$ForLoadedMethod.getDeclaringType(MethodDescription.java:1190)
	at org.mockito.internal.creation.bytebuddy.MockMethodAdvice.isOverridden(MockMethodAdvice.java:199)
	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.isUnavailable(ConsumerNetworkClient.java:560)
	at org.apache.kafka.clients.consumer.internals.Fetcher.isUnavailable(Fetcher.java:87)
	at org.apache.kafka.clients.consumer.internals.AbstractFetch.prepareFetchRequests(AbstractFetch.java:427)
	at org.apache.kafka.clients.consumer.internals.Fetcher.sendFetches(Fetcher.java:105)
	at org.apache.kafka.clients.consumer.internals.FetcherTest.sendFetches(FetcherTest.java:246)
	at org.apache.kafka.clients.consumer.internals.FetcherTest.testFetcherConcurrency(FetcherTest.java:2943)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:728)
```

the failed test is related to this PR. In the test case `testFetcherConcurrency`, it does not return correct `sessionTopicNames` so normally it should NOT see the correct response data. However, `FetchResponse#responseData` will return the cached data regardless of input, so it CAN get correct response data even though it pass empty `topicNames`. That is a good example of showing the potential bug :)

@m1a2st Could you copy the changes of this PR to another one, and please fix `testFetcherConcurrency` according to my comment. Also, please add new test for the change.

@johnnychhsu Sorry, I can't merge this PR as it causes the failed test. Please feel free to close this PR as @m1a2st will leverage this PR to complete it :)

**Comment by m1a2st** [comment]
@chia7712, Thanks for your comments, I will open new PR for this issue.

**Comment by chia7712** [comment]
close this PR @m1a2st  will file another one

**Pull request closed without merging**
